### PR TITLE
[MSAFD][AFD][TCP] Add asynchronous connection support

### PR DESCRIPTION
**Pull request by mann3616**
## Purpose

This is a PR of a combined patch of patches that was fixed by Oleg Dubinskiy and originally made by hater (on JIRA) which adds asynchronous connection support (thereby allowing torrent programs to work, web browsing to be faster, etc).
JIRA issue: [CORE-12152](https://jira.reactos.org/browse/CORE-12152)

## Proposed changes
On top of the previous patch by the above devs. I added a line in WSPConnect which would allow WSPCloseSocket to identify that the socket is connected so that it can allow for the AfdConnectAPC to run first.

### CODE REVIEW

File: drivers/network/tcpip/ip/transport/tcp/tcp.c [hunk 1]
```diff
@@ -365,8 +365,12 @@ NTSTATUS TCPConnect
...
     }
 
-    /* Copy bind address into connection */
-    Connection->AddressFile->Address.Address.IPv4Address = bindaddr.addr;
+    /* Check if we had an unspecified address */
+    if (Connection->AddressFile->Address.Address.IPv4Address != bindaddr.addr)
+    {
+         /* We did, so we need to copy back the address */
+         Connection->AddressFile->Address.Address.IPv4Address = bindaddr.addr;
+    }
```
Comment (HBelusca) [inline drivers/network/tcpip/ip/transport/tcp/tcp.c:373]: Why can't we just always copy back the address?

File: drivers/network/afd/afd/write.c [hunk 2]
```diff
@@ -656,7 +664,7 @@ AfdPacketSocketWriteData(PDEVICE_OBJECT DeviceObject, PIRP Irp,
             {
                 NT_VERIFY(RemoveHeadList(&FCB->PendingIrpList[FUNCTION_SEND]) == &Irp->Tail.Overlay.ListEntry);
                 Irp->IoStatus.Status = Status;
-                Irp->IoStatus.Information = 0;
+                Irp->IoStatus.Information = Status == STATUS_SUCCESS ? SendReq->BufferArray[0].len : 0;
```
Comment (HBelusca) [inline drivers/network/afd/afd/write.c:667]:
```suggestion
                Irp->IoStatus.Information = ((Status == STATUS_SUCCESS) ? SendReq->BufferArray[0].len : 0);
```

File: drivers/network/afd/afd/write.c [hunk 3]
```diff
@@ -390,7 +394,7 @@ AfdConnectedSocketWriteData(PDEVICE_OBJECT DeviceObject, PIRP Irp,
                 {
                     NT_VERIFY(RemoveHeadList(&FCB->PendingIrpList[FUNCTION_SEND]) == &Irp->Tail.Overlay.ListEntry);
                     Irp->IoStatus.Status = Status;
-                    Irp->IoStatus.Information = 0;
+                    Irp->IoStatus.Information = Status == STATUS_SUCCESS ? SendReq->BufferArray[0].len : 0;
```
Comment (HBelusca) [inline drivers/network/afd/afd/write.c:397]:
```suggestion
                    Irp->IoStatus.Information = ((Status == STATUS_SUCCESS) ? SendReq->BufferArray[0].len : 0);
```

File: dll/win32/msafd/misc/dllmain.c [hunk 4]
```diff
@@ -1821,6 +1821,49 @@ WSPAccept(
...
+        Context->lpSocket->SharedData->ConnectTime =
+            (GetTickCount() - Context->lpSocket->SharedData->ConnectTime) / 1000;
+        Context->lpSocket->SharedData->State = SocketConnected;
+        Context->lpSocket->TdiConnectionHandle = (HANDLE)IoStatusBlock->Information;
+    }
+
+    /* Re-enable Async Event */
+    SockReenableAsyncSelectEvent(Context->lpSocket, FD_WRITE);
+
+    /* FIXME: THIS IS NOT RIGHT!!! HACK HACK HACK! */
```
Comment (HBelusca) [inline dll/win32/msafd/misc/dllmain.c:1849]: Any explanation (in code comment) for what needs to be done instead?

File: dll/win32/msafd/misc/dllmain.c [hunk 5]
```diff
@@ -1970,63 +2001,76 @@ WSPConnect(SOCKET Handle,
     ConnectInfo->UseSAN = FALSE;
     ConnectInfo->Unknown = 0;
 
-    /* FIXME: Handle Async Connect */
     if (Socket->SharedData->NonBlocking)
     {
-        ERR("Async Connect UNIMPLEMENTED!\n");
+        APCFunction = &AfdConnectAPC; // should be a private io completition function inside us
```
Comment (HBelusca) [inline dll/win32/msafd/misc/dllmain.c:2006]: Is it understood to be a future FIXME?

File: drivers/network/tcpip/ip/transport/tcp/tcp.c (hunk 1 above)
Comment (mann3616) [reply drivers/network/tcpip/ip/transport/tcp/tcp.c:373]: Only change I made was commit [f5f4740](https://github.com/reactos/reactos/pull/5092/commits/f5f47402da15870c579a02c3dc8eea3c0d4987f2)
Sorry I am trying to clean up the commit history right now. The other changes were made by hater.

Comment (mann3616) [reply drivers/network/tcpip/ip/transport/tcp/tcp.c:373]: Commit history is fixed, the first commit [0d9ab98](https://github.com/reactos/reactos/pull/5092/commits/0d9ab989975193f597dace30610217503684adcf) and mine is [f5f4740](https://github.com/reactos/reactos/pull/5092/commits/f5f47402da15870c579a02c3dc8eea3c0d4987f2)

**Comment by JoachimHenze** [comment]
It is known, that the hater async patch is not side-effect-free yet regarding reliability of DNS-lookups. But I think ros is currently stable enough to survive a bit of added instability here, in order to move ahead. I still want to see testbots results of this PR before it being merged.

**Comment by mann3616** [comment]
On top of this, combined with the [LDNS PR](https://github.com/reactos/reactos/pull/5018), async no longer has the DNS lookup issue.

**Comment by KRosUser2** [comment]
Any compatibility issue with @GGLinnk PR on new LwIP ?

File: dll/win32/msafd/misc/dllmain.c [hunk 6]
```diff
@@ -3500,11 +3545,10 @@ SetSocketInformation(PSOCKET_INFORMATION Socket,
     if (Status == STATUS_PENDING)
     {
         TRACE("Leaving (Pending)\n");
-        return WSA_IO_PENDING;
+        // return MsafdReturnWithErrno(Status, NULL, IOSB->Information, NULL);
```
Comment (ThFabba) [inline dll/win32/msafd/misc/dllmain.c:3548]: I think you can just remove this block, the "Pending" trace adds no additional information compared to the Status/Information one.

File: dll/win32/msafd/misc/sndrcv.c [hunk 7]
```diff
@@ -166,9 +196,9 @@ WSPRecv(SOCKET Handle,
 {
     PIO_STATUS_BLOCK        IOSB;
     IO_STATUS_BLOCK         DummyIOSB;
-    AFD_RECV_INFO           RecvInfo;
+    PAFD_RECV_INFO          RecvInfo;
```
Comment (sdever) [inline dll/win32/msafd/misc/sndrcv.c:199]: Why switching from a local variable to a heap-allocated variable that gets freed at the end of the function ?

**Comment by julenuri** [comment]
As the MSAFD code changed, I had to redo the patch according to master. Anyway, could have mistakes and whitespaces and be wrong, but don't fix my testing situation. 
[MSAFD-5092.txt](https://github.com/reactos/reactos/files/13258618/MSAFD-5092.txt)

**Pull request closed without merging**